docs/proposals/network-policy.md

```diff
@@ -0,0 +1,174 @@
+
+Each ingress rule supports the following fields:
+- `ports`: A list of protocol / ports which should be accessible on the selected pods. If not defined, then access to the selected pods will not be restricted by port/protocol.
+- `from`: A list of source criteria which dictate which sources should be able to access the pods selected. If no `from` key is provided, then access will not be restricted based on source.
+
+Each item in the `ports` list supports the following fields:
+- `protocol`: The protocol for the given port - TCP or UDP.
+- `port`: The port on the specified protocol.
+
+Each item in the `from` list may declare at most one of the following:
+- `pods`: A label selector which selects pods (in the same namespace as the NetworkPolicy).
+- `namespaces`: A label selector which selects namespaces.
+
+Note:
+- Any traffic which does not match one of the NetworkPolicy ingress rules will be dropped via a default
+drop behavior (this occurs whether or not network-isolation has been enabled on the Namespace).
+- All ingress rules are whitelist rules, meaning that it should be easy to resolve the case where multiple
+NetworkPolicy objects select the same set of pods, as there can be no conflicting rules.
+- All pods will always be accessible from the host that they are running on.  This is required to allow for kubelet health checks.
+
+### Go Representation
+
+```go
+type NetworkPolicy struct {
+  TypeMeta
+  ObjectMeta
+
+  Spec NetworkPolicySpec 
+}
+
+type NetworkPolicySpec struct {
+  // Selects the pods which this NetworkPolicy object
+  // applies to.
+  PodSelector *unversioned.LabelSelector `json:"podSelector,omitempty"`
+
+  // List of ingress rules to be applied to the 
+  // selected pods.
+  Ingress []NetworkPolicyIngressRule `json:"ingress,omitempty"`
+}
+
+type NetworkPolicyIngressRule struct {
+  // List of allowed ports. 
+  Ports []NetworkPolicyPort `json:"ports,omitempty"`
+
+  // List of allowed sources.
+  From []NetworkPolicySource `json:"from,omitempty"`
+}
+
+type NetworkPolicyPort struct {
+  // If specified, the port on the given protocol.
+  Port int32 `json:"port,omitempty"`
+
+  // The protocol - TCP or UDP.
+  Protocol string `json:"protocol"`
+}
+
+type NetworkPolicySource struct {
+  // Label selector - selects pods in this namespace.
+  // If 'Namespaces' is defined, this must not be.
+  Pods *unversioned.LabelSelector `json:"pods,omitempty"`
+
+  // Label selector - selects namespaces.
+  // If 'Pods' is defined, this must not be.
+  Namespaces *unversioned.LabelSelector `json:"namespaces,omitempty"`
+}
+```
+
+
+## References
+
+- https://github.com/kubernetes/kubernetes/issues/22469 tracks network policy in kubernetes.
+
+
+<!-- BEGIN MUNGE: GENERATED_ANALYTICS -->
+[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/docs/proposals/network-policy.md?pixel)]()
+<!-- END MUNGE: GENERATED_ANALYTICS -->
```

Comment:
Is this document the right place for describing validation constraints on policy objects? (e.g.: either 'pod' or 'namespace' selector in any 'from' clause, but not both)